File: back/src/main/java/com/back/domain/tag/service/TagService.java
package com.back.domain.tag.service;

import com.back.domain.tag.entity.Tag;
import com.back.domain.tag.repository.TagRepository;
import com.back.global.exception.ServiceException;
import jakarta.validation.constraints.NotBlank;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Optional;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class TagService {

    private final TagRepository tagRepository;

    @Transactional
    public void post(String name) {

        Optional<Tag> tag = tagRepository.findByName(name);

        if (tag.isPresent()) {
            throw new ServiceException("409-1", "이미 존재하는 태그입니다.");
        }

        tagRepository.save(new Tag(name));
    }

    @Transactional
    public Tag findByNameOrSave(String name) {
        return tagRepository.findByName(name)
                .orElseGet(() -> tagRepository.save(new Tag(name)));
    }
}

File: back/src/main/kotlin/com/back/domain/tag/service/TagService.kt
package com.back.domain.tag.service

import com.back.domain.tag.entity.Tag
import com.back.domain.tag.repository.TagRepository
import com.back.global.exception.ServiceException
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true)
class TagService(
    private val tagRepository: TagRepository
) {

    @Transactional
    fun createTag(name: String) {
        if (tagRepository.findByName(name) != null)
            throw ServiceException("409-1", "이미 존재하는 태그입니다.")

        tagRepository.save(Tag(name))
    }

    @Transactional
    fun findByNameOrSave(name: String): Tag =
        tagRepository.findByName(name) ?: tagRepository.save(Tag(name))
}
